### PR TITLE
Fix duplicate file paths in error reporting

### PR DESCRIPTION
**Pull request by tosky1125**
## Description

This PR fixes the issue where the same file is reported multiple times in error output when accessed through different path representations (e.g., `doc/docker.md` vs `./doc/docker.md`).

Fixes #[issue-number]

## Problem

When `lychee` discovers files through different methods (CLI arguments vs document links), the same file can be represented with different path formats:
- `doc/docker.md` (from glob pattern)
- `./doc/docker.md` (from document link)

This causes duplicate entries in the error report, making it appear as if the same error occurred in multiple files.

## Solution

Implemented path normalization in the `Display` trait for `InputSource::FsPath` to ensure consistent path representation across all platforms:

- Removes redundant `./` prefixes (including multiple occurrences like `./././`)
- Filters out intermediate `.` components (e.g., `doc/./file.md` → `doc/file.md`)
- Converts Windows backslashes to forward slashes for cross-platform consistency
- Preserves important path characteristics (absolute paths, parent directory references, trailing slashes)
- Handles edge cases like empty paths and root directories

## Changes

- Modified `Display` implementation for `InputSource` in `lychee-lib/src/types/input.rs`
- Added comprehensive test coverage for various path formats and edge cases

## Testing

Added extensive tests covering:
- [x] Basic path normalization (`./doc/file.md` → `doc/file.md`)
- [x] Multiple current directory markers (`./././doc/file.md` → `doc/file.md`)
- [x] Windows path normalization (`doc\file.md` → `doc/file.md`)
- [x] Cross-platform consistency
- [x] Edge cases (empty paths, trailing slashes, absolute paths)
- [x] All existing tests pass

## Notes

- The normalization only affects display output, not the actual file processing logic
- Maintains backward compatibility with existing behavior
- No performance impact as normalization only occurs during error reporting

### CODE REVIEW

**Comment by tosky1125** [comment]
The test failures appear to be unrelated to this PR's changes. The failing tests are:
- `test_formatted_file_output`
- `test_exclude_multiple_urls`

Both are failing due to `https://creativecommons.org/publicdomain/zero/1.0/` returning 403 Forbidden, which seems to be an external service issue or rate limiting in the CI environment.

The changes in this PR only affect path display normalization and all path-related tests pass successfully.

**Review by mre — COMMENTED** [review]
Interesting idea. 

- The issue name is not filled out in the pull request description. 
- There are a few comments in the PR code, which I would clean up. Will highlight them. 
- Needs an Integration test in cli.rs
- Does Windows path conversion in multiple places. I'm also not sure if there aren't any edge cases with that. Maybe there's a library for it. Also wondering if we even need to change the Windows paths because I think it's good to know the exact path that was checked, even if that means the output is different on different platforms. 
- I think the path cleanup should be in a separate method (or even module). Display should not have a lot of business logic, but just call into other methods if needed.

File: lychee-lib/src/types/input.rs [hunk 1]
```diff
@@ -90,13 +90,57 @@ impl Serialize for InputSource {
...
 impl Display for InputSource {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str(match self {
-            Self::RemoteUrl(url) => url.as_str(),
-            Self::FsGlob { pattern, .. } => pattern,
-            Self::FsPath(path) => path.to_str().unwrap_or_default(),
-            Self::Stdin => "stdin",
-            Self::String(s) => s,
-        })
+        use std::path::{Component, PathBuf}; // Move to top of function
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: Guess this can be cleaned up?

File: lychee-lib/src/types/input.rs [hunk 2]
```diff
@@ -90,13 +90,57 @@ impl Serialize for InputSource {
...
+
+        match self {
+            Self::RemoteUrl(url) => f.write_str(url.as_str()),
+            Self::FsGlob { pattern, .. } => f.write_str(pattern),
+            Self::FsPath(path) => {
+                // Use to_string_lossy() for safety with non-UTF8 paths
+                let original = path.to_string_lossy();
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: You're only using `original` later. I would move it further down.

File: lychee-lib/src/types/input.rs [hunk 3]
```diff
@@ -90,13 +90,57 @@ impl Serialize for InputSource {
...
+
+                let mut result = normalized.to_string_lossy().into_owned();
+
+                // Handle empty path
+                if result.is_empty() {
+                    result = ".".to_string();
+                }
+
+                // Preserve trailing slash if present in original
+                // Fix collapsible_if warning
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: Redundant comment

File: lychee-lib/src/types/input.rs [hunk 4]
```diff
@@ -90,13 +90,57 @@ impl Serialize for InputSource {
...
+                    return f.write_str(&original);
+                }
+
+                // For relative paths, do component-based normalization
+                let normalized: PathBuf = path
+                    .components()
+                    .filter(|comp| !matches!(comp, Component::CurDir))
+                    .collect();
+
+                let mut result = normalized.to_string_lossy().into_owned();
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: Looks weird. I wonder if we can avoid the string conversion or at least the `into_owned`.

File: lychee-lib/src/types/input.rs [hunk 5]
```diff
@@ -90,13 +90,57 @@ impl Serialize for InputSource {
...
+                    && !result.ends_with('/')
+                    && result != "."
+                {
+                    result.push('/');
+                }
+
+                // Ensure forward slashes on Windows
+                #[cfg(windows)]
+                {
+                    result = result.replace('\\', "/");
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: This is the same code as above. I also don't want to litter the code with windows checks. I would see if there's a method in the standard library or a crate in the Rust, which handles the path conversion. My hunch is that it's not that easy.

File: lychee-lib/src/types/input.rs [hunk 6]
```diff
@@ -496,6 +540,124 @@ mod tests {
...
+    fn test_normalized_path_display() {
+        // Test that paths with and without "./" prefix are displayed identically
+        let path1 = InputSource::FsPath(PathBuf::from("doc/docker.md"));
+        let path2 = InputSource::FsPath(PathBuf::from("./doc/docker.md"));
+        assert_eq!(path1.to_string(), path2.to_string());
+        assert_eq!(path2.to_string(), "doc/docker.md");
+    }
+
+    #[test]
+    fn test_path_display_comprehensive() {
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: We should use the rstest crate for this.

File: lychee-lib/src/types/input.rs [hunk 7]
```diff
@@ -496,6 +540,124 @@ mod tests {
...
+            ("doc\\sub/file.md", "doc/sub/file.md"),
+            (".\\doc/file.md", "doc/file.md"),
+            // Windows absolute paths
+            ("C:\\Users\\file.md", "C:/Users/file.md"),
+            (
+                "C:\\Users\\Documents\\file.md",
+                "C:/Users/Documents/file.md",
+            ),
+            // UNC paths
+            ("\\\\server\\share\\file.md", "//server/share/file.md"),
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: This is strange. I was not expecting that it changes the `\\\\` to `//`. Could potentially be confused with a different path, but I don't have an example right now. Not sure.

Comment (mre) [reply lychee-lib/src/types/input.rs]: If anything, why not just a single slash in the beginning? Not sure how UNC paths work.

File: lychee-lib/src/types/input.rs [hunk 8]
```diff
@@ -496,6 +540,124 @@ mod tests {
...
+        ];
+
+        for (input, expected) in windows_cases {
+            let source = InputSource::FsPath(PathBuf::from(input));
+            assert_eq!(
+                source.to_string(),
+                expected,
+                "Failed for Windows input: '{}'",
+                input
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: This input variable could be inlined into the format string.

**Comment by tosky1125** [comment]
@mre Thank you for the thorough review! I've addressed most of your feedback:

### Changes made:
1.  **Extracted normalization logic**: Moved path normalization from `Display` impl to a dedicated `InputSource::normalize_path_display()` method for better separation of concerns.

2. **Added integration test**: Added `test_normalized_path_display` in `cli.rs` to verify that paths like `./test.md` are displayed consistently as `test.md`.

3. **Removed Windows path conversion**: As you suggested, I've removed the backslash-to-forward-slash conversion to preserve platform-specific paths. Now only `.` normalization is performed.

4. **Cleaned up comments**: Removed redundant comments and moved variable declarations closer to their usage.

5. **Simplified implementation**: The current implementation only removes leading `./` for consistency, avoiding complex path manipulations.

Additionally, I've confirmed that the CI failures are unrelated to this PR:

The failing tests expect:
- `successful: 10`
- `errors: 0`

But actually get:
- `successful: 7` 
- `errors: 1` (403 from creativecommons.org)

Our path normalization only affects relative paths (removes `./` prefix) and preserves absolute paths unchanged. The JSON serialization works correctly with our changes.

This is purely an external service issue affecting all PRs.

File: lychee-lib/src/types/input.rs [hunk 9]
```diff
@@ -88,15 +88,50 @@ impl Serialize for InputSource {
...
+            ".".to_string()
+        } else {
+            cleaned.to_string_lossy().into_owned()
+        };
+
+        // Preserve trailing slash for "./"
+        if path_str == "./" {
+            "./".to_string()
+        } else if has_trailing_slash && result != "." && !result.ends_with('/') {
+            format!("{result}/") // 수정된 부분
```
Comment (mre) [inline lychee-lib/src/types/input.rs]: Comment can be removed

Comment (tosky1125) [reply lychee-lib/src/types/input.rs]: removed 👍

**Comment by thomas-zahner** [comment]
Thank you for the PR. This is an interesting idea. However, this does not really fix https://github.com/lycheeverse/lychee/issues/1660 as to my understanding. This now "normalises" output files to some specific format. But the affected files are still checked multiple times and also displayed multiple times. To really address the issue we would need to filter out such duplicates before checking them.

File: lychee-bin/tests/cli.rs [hunk 10]
```diff
@@ -2222,4 +2222,24 @@ mod cli {
...
+
+        let mut cmd = main_command();
+        cmd.arg("test.md").arg("./test.md").current_dir(dir.path());
+
+        let output = cmd.output()?;
+        let stdout = String::from_utf8_lossy(&output.stdout);
+
+        // Both should be displayed as "test.md" (not "./test.md")
+        assert!(stdout.contains("[test.md]:"));
+        assert!(!stdout.contains("[./test.md]:"));
```
Comment (thomas-zahner) [inline lychee-bin/tests/cli.rs:2241]: To really fix https://github.com/lycheeverse/lychee/issues/1660 as to my understanding the assertions must check that `test.md` (or `./test.md`) occurs exactly once in `stdout`. In the current state of the PR this assertion will fail because `test.md` occurs twice.

**Comment by mre** [comment]
> this does not really fix https://github.com/lycheeverse/lychee/issues/1660 

That is correct. @tosky1125, can you remove the `Fixes #1660` from the description? I think the PR is still valuable to be merged as is because it cleans up the output a bit. @thomas-zahner, do you agree?

File: lychee-bin/600c0d6d-eacd-4fba-bef0-bd79a4da72c2.json [hunk 11]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (thomas-zahner) [inline lychee-bin/600c0d6d-eacd-4fba-bef0-bd79a4da72c2.json:1]: What's the purpose of all these JSON files? They don't seem to be used in the test, right?

File: lychee-lib/src/types/input.rs [hunk 12]
```diff
@@ -88,15 +88,50 @@ impl Serialize for InputSource {
...
 impl Display for InputSource {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str(match self {
-            Self::RemoteUrl(url) => url.as_str(),
-            Self::FsGlob { pattern, .. } => pattern,
-            Self::FsPath(path) => path.to_str().unwrap_or_default(),
-            Self::Stdin => "stdin",
-            Self::String(s) => s,
-        })
+        match self {
```
Comment (thomas-zahner) [inline lychee-lib/src/types/input.rs:128]: I'd much prefer the previous position of `match` as this didn't duplicate `f.write_str` five times

**Review by thomas-zahner — COMMENTED** [review]
After giving it some thought, I came to the conclusion that I would prefer not to merge this. As mentioned earlier addressing the root cause to eliminate duplicates would be great, but this PR does not solve that. 

This PR applies some cosmetic changes to the output. However, I actually prefer the current output because it contains more information.

Given the following example `cargo run *.md ./*.md --dump-inputs` we currently get

```
CONTRIBUTING.md
README.md
./CONTRIBUTING.md
./README.md
```

with this PR I would expect the following (noticed that it's not yet the case because this PR only affects normal link checking as of now, not dumping [which would be another reason for not merging the PR because I expect consistent behaviour as a user across all commands])

```
CONTRIBUTING.md
README.md
CONTRIBUTING.md
README.md
```

This removes some useful information in my opinion. I no longer understand as easily why my files were included twice for link checking.

**Comment by mre** [comment]
I went ahead and closed this as it doesn't resolve an immediate problem. Thanks anyway for the idea and the PR, @tosky1125.

File: lychee-bin/600c0d6d-eacd-4fba-bef0-bd79a4da72c2.json (hunk 11 above)
Comment (thomas-zahner) [reply lychee-bin/600c0d6d-eacd-4fba-bef0-bd79a4da72c2.json:1]: Realised that these JSON files were likely created with the `test_json_output` macro after test failures. This was because the files were not cleaned up after a failing assertion. I've fixed that behaviour with https://github.com/lycheeverse/lychee/pull/1771/commits/ec4bedb3d4d21ecc0a2db9b6f94423d4f43ca8ba

**Comment by thomas-zahner** [comment]
Yes, thank you nonetheless for the idea and your efforts @tosky1125.
